src/clients/c++/examples/simple_cuda_shm_client.cc

```diff
@@ -0,0 +1,497 @@
+              << std::endl;
+    exit(1);
+  }
+  // extend shared memory object as by default it's initialized with size 0
+  int res = ftruncate(shm_fd, batch_byte_size);
+  if (res == -1) {
+    std::cerr << "error: unable to initialize shared-memory key '" + shm_key +
+                     "' to requested size " + std::to_string(batch_byte_size) +
+                     " bytes"
+              << std::endl;
+    exit(1);
+  }
+  return shm_fd;
+}
+
+void*
+MapSharedMemory(int shm_fd, size_t offset, size_t batch_byte_size)
+{
+  // map shared memory to process address space
+  void* shm_addr =
+      mmap(NULL, batch_byte_size, PROT_WRITE, MAP_SHARED, shm_fd, offset);
+  if (shm_addr == MAP_FAILED) {
+    std::cerr << "error: unable to process address space or shared-memory "
+                 "descriptor: " +
+                     std::to_string(shm_fd)
+              << std::endl;
+    exit(1);
+  }
+  return shm_addr;
+}
+
+void
+UnlinkSharedMemoryRegion(std::string shm_key)
+{
+  int shm_fd = shm_unlink(shm_key.c_str());
+  if (shm_fd == -1) {
+    std::cerr << "error: unable to unlink shared memory for key '" + shm_key +
+                     "'"
+              << std::endl;
+    exit(1);
+  }
+}
+
+void
+UnmapSharedMemory(void* shm_addr, size_t byte_size)
+{
+  int tmp_fd = munmap(shm_addr, byte_size);
+  if (tmp_fd == -1) {
+    std::cerr << "Unable to munmap shared memory region" << std::endl;
+    exit(1);
+  }
+}
+
+void
+CreateCUDAIPCHandle(ipcCUDA_t* shm_cuda_rep, int* input_d_ptr)
+{
+  shm_cuda_rep->device = 0;
+  checkCudaErrors(cudaSetDevice(shm_cuda_rep->device));
+
+  cudaEvent_t event;
+
+  //  allocate data on gpu to ipc handle, create an event
+  checkCudaErrors(
+      cudaEventCreate(&event, cudaEventDisableTiming | cudaEventInterprocess));
+  checkCudaErrors(cudaIpcGetEventHandle(
+      (cudaIpcEventHandle_t*)&shm_cuda_rep->eventHandle, event));
+  cudaIpcOpenMemHandle(
+      (void**)&d_ptr, *(cudaIpcMemHandle_t*)&shm_cuda_rep->memHandle,
+      cudaIpcMemLazyEnablePeerAccess)
+
+      // set device to default GPU - GPU0
+      checkCudaErrors(cudaSetDevice(DEVICE_ID));
+}
+
+void
+ReadCUDAIPCHandle(ipcCUDA_t* shm_cuda_rep, int* output_data)
+{
+  checkCudaErrors(cudaSetDevice(shm_cuda_rep->device));
+
+  int* d_ptr;
+  cudaEvent_t event;
+
+  // get cuda event and synchronize
+  checkCudaErrors(cudaIpcOpenEventHandle(&event, shm_cuda_rep->eventHandle));
+  checkCudaErrors(cudaEventSynchronize(event));
+
+  // allocate data on the gpu and copy out output data
+  checkCudaErrors(cudaMalloc((void**)&d_ptr, 16 * sizeof(int)));
+  checkCudaErrors(cudaIpcGetMemHandle(
+      (cudaIpcMemHandle_t*)&shm_cuda_rep->memHandle, (void*)d_ptr));
+  checkCudaErrors(cudaMemcpy(
+      (void*)output_data, (void*)d_ptr, 16 * sizeof(int),
+      cudaMemcpyDeviceToHost));
+
+  // close IPC handle and free output data on GPU
+  checkCudaErrors(cudaIpcCloseMemHandle(d_ptr));
+  checkCudaErrors(cudaFree(d_ptr));
```

Comment:
Either change the function name to suggest that the handle will also be closed, or move this out to where it closes the input handles

Reply:
Moving to where input handles are destoryed